palantirnet/federated-search-react
# Root 198: add search-as-you-type autocomplete functionality option to text search

This PR adds the option of enabling a configuring the main query text field to implement search as you type functionality, taking advantage of new functionality built in https://github.com/palantirnet/solr-faceted-search-react/pull/7

## To test

### Get a local instance of the federated search demo up and running
1. Check out master in the [Drupal federated-search-demo](https://github.com/palantirnet/federated-search-demo) project
1. Boot the vm `vagrant up`, ssh in to the vm `vagrant ssh` and [build the sites](https://github.com/palantirnet/federated-search-demo#rebuild-all-the-things) with `phing build-all init`
1. Log in to the standalone d8 site `drush @fsd-d8.local uli`
1. Clear + reindex all items in the index at: http://d8.fs-demo.local/admin/config/search/search-api/index/federated_search_index
1. Note: we'll be using the index provided by the demo site, but won't yet be using the search page on the demo site

### Get your local app environment set up
1. Pull this branch down
1. Update [the version of the search app dependency](https://github.com/palantirnet/solr-faceted-search-react/pull/7) and install some new dependencies with `yarn install`

### Verify when no autocomplete configuration present, the app functions as it used to, with a standard text search functionality
1. Start the app with `yarn start`
1. The search app should load in a new browser window
1. Observe you can still search and use the facets/filters

### Add new autocomplete config
1. Observe updates to the `src/env.local.js.example` file
1. Add configuration for your solr backend and autocomplete to your `src/env.local.js`, here is what mine looks like:

```js
module.exports = {
  // REQUIRED: The default solr backend.
  // url: "https://ss826806-us-east-1-aws.measuredsearch.com:443/solr/master/select",
  url: "http://federated-search-demo.local:8983/solr/drupal8/select",
  // OPTIONAL: The default "Site Name" facet value.
  // Note: This value must match the "site_name" property for one of your sites.
  // siteSearch: 'U of M Lab Blog',
  // userpass: btoa("palantir:palantirqauser"),
  // OPTIONAL: The text to display when a search returns no results.
  // noResults: "Sorry, your search yielded no results.",
  // OPTIONAL: The text to display when the app loads with no search term.
  searchPrompt: "Please enter a search term.",
  // OPTIONAL: The number of search results to show per page.
  rows: 20,
  // OPTIONAL: The number of page buttons to show for pagination.
  paginationButtons: 5,
  // OPTIONAL: Provides config for adding autocomplete functionality to text search
  autocomplete: {
    mode: 'result',
    url: 'http://federated-search-demo.local:8983/solr/drupal8/select',
    queryField: 'tem_suggestion_title',
    suggestionRows: 5,
	numChars: 2,
    result: {
      titleText: 'What are you interested in?',
      directionsText: true,
    },
  }
};
```

### Test basic autocomplete functionality
1. Upon saving your `src/env.local.js` file, observe the app reloads and that the text search input looks identical
1. Begin typing search queries and observe that after the 2nd character (or whatever value you used for `autocomplete.numChars` in your `src/env.local.js` file) the search as you type suggestion results render.
1. Observe that as you continue to type, the results get refined.
1. Observe that, when the the suggestions render:
    1. there are instructions for sighted keyboard users
    1. there is a close button to clear suggestions

### Test that facet / filter values are honored in suggestion query
1. Search for "uma"
1. Observe the suggestions (and observe Uma and uma are highlighted in suggestion text)
1. Expand the `Type` filter accordion
1. Select `Page`
1. Bring focus to the search input by clicking inside of it
1. Observe the suggestions now reflect the `Type=Page` filter

## Next steps
- ~remove any default field that is currently sent in the query~
- ~ensure that query parsers, fields (including defaults and boosts) can be configured in solr config~
- ~add the ability to hide any facet / filter group from app config~
- update the module app config to pass in:
	- autocomplete
	- hidden facet/filter groups
	- more default values?
